### ISSUE TITLE
[Suggestion] Remove/relax CS1738 "Named argument specifications must appear after all fixed arguments have been specified"

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#518]
@dsaf commented on [Thu Jun 02 2016](https://github.com/dotnet/roslyn/issues/11713)

I like using named arguments for code readability when passing literals. However the subj limitation forces me to either not use them or specify them for every subsequent argument:

``` csharp
var name = "John";
var age = 30;

DoSomething(isEmployed:true, name, age); //CS1738
```

---

@bondsbw commented on [Mon Jun 06 2016](https://github.com/dotnet/roslyn/issues/11713#issuecomment-224130781)

I like this, but only if argument order aligns with the parameter list until at least the last unnamed argument.

Example:

``` c#
public void DoSomething(bool isEmployed, string personName, int personAge) { ... }

DoSomething(true, personName:name, age); // would be legal
DoSomething(name, isEmployed:true, age); // remains illegal
DoSomething(name, age, isEmployed:true); // remains illegal
DoSomething(true, personAge:age, personName:name); // already legal
```

---

@dsaf commented on [Tue Jun 07 2016](https://github.com/dotnet/roslyn/issues/11713#issuecomment-224219177)

@bondsbw of course.

**Comment by jnm2** [comment]
Yes! Another frequent wish.

**Comment by sharwell** [comment]
@gafter I would be interested in your feedback regarding the best way to move this proposal forward.

The current language specification contains the following limitation:

> It is an error for a positional argument to appear after a named argument in an *argument_list*.

I would suggest the following alternative:

> It is an error for a positional argument to appear after a named argument in an *argument_list* for which the named argument's position does not equal the corresponding parameter position in the target of the invocation.

This change in wording should not negatively impact the ability of named arguments to be used for overload resolution, even in cases like the following:

```csharp
private static void Foo(int x, int a) { }
private static void Foo<T>(T y, int a) { }

private static void Bar()
{
    Foo(3, 4);
    Foo(y: 3, 4);
}
```

I am worried about the complexity this feature would introduce in dealing with the following:

```csharp
private static void Foo(int x, int y) { }
private static void Foo<T>(T y, int x) { }

private static void Bar()
{
    Foo(3, 4);
    Foo(y: 3, 4);
}
```

**Comment by gafter** [comment]
> @gafter I would be interested in your feedback regarding the best way to move this proposal forward.

@sharwell The next step would be to find someone on the LDM @dotnet/csharplangdesign who believes this is both desirable and of higher priority than many other things we could do, and who is willing to invest their time to push the issue forward. That LDM member would then volunteer to "champion" this proposal, and bring it to LDM to be triaged. I believe there are two or three members of the LDM on your team.

**Comment by gafter** [comment]
@sharwell Also, one should compare doing this language feature with the alternative of doing an IDE feature that fills in the parameter for you at the call site.

**Comment by DavidArno** [comment]
I think it reasonable to assume that the overwhelmingly most common use-case for this will be for handling `bool` parameters:
```cs
 DoSomething(isEmployed:true, name, age);
```
The use of bool parameters like this is a code smell in my view, as it's very likely that `DoSomething` will like:
```cs
void DoSomething(bool isEmployed, string Name, int Age)
{
    if (isEmployed)
    {
        // do something with employed person
    }
    else
    {
        // do something with unemployed person
    }
}
```
In such situations, it makes more sense to provide two methods:
```cs
DoSomethingWhenEmployed(name, age);
DoSomethingWhenUnemployed(name, age);
```

**Comment by bondsbw** [comment]
But then you double each overload that includes the bool.  And more bools causes them to increase exponentially.

The main reason bools are used is laziness.  A new enum requires typing more characters and LOC than the `bool` keyword, and the `EnumeratedType.Option` callsite syntax is verbose. `using static` makes this much nicer.

ADTs would be even better, preventing the use of invalid values.

**Comment by sharwell** [comment]
@DavidArno passing `null` is another common case. And while you could argue for your own code to avoid cases where this would be relevant, it's a moot point for published third-party libraries with stable APIs. Maybe it should have been different, but it won't/can't be now.

Literals of various kinds are also frequent in unit tests and named parameters can help readers be confident that the correct methods are being tested.

**Comment by DavidArno** [comment]
@sharwell,
> ... it's a moot point for published third-party libraries with stable APIs. Maybe it should have been different, but it won't/can't be now.

I have a real problem with this stance. I appreciate that "never make a breaking change" is deeply embedded in Microsoft traditions, but that is no excuse to deny there is an alternative. Semantic versioning is specifically designed to handle making such changes to libraries. The approach of v2.0.0 depreciating an API then v3.0.0 removing it completely works very well. So I disagree: it's not a moot point at all.

**Comment by bondsbw** [comment]
Like most things regarding BC, breaking it should probably be left to an analyzer.  But this proposal is a good start and can even be coupled with an IDE feature that targets literal arguments.

**Comment by jnm2** [comment]
I have this problem with ints too, not just bools and null.

**Comment by mattwar** [comment]
Yes, any time you are passing a literal value where it is not clear what the value represents it is useful to use a named parameter to help document what is happening in the code, but a nuisance that all following parameter must now be named as well.

**Comment by sharwell** [comment]
@gafter I believe the main value in this comes from the fact that the place I see this requested the most is in code reviews, where IDE features aren't playing a role. Perhaps we can get GitHub someday to incorporate Roslyn into their PR UI (can you imagine?!), but until then I have to say this is a pretty compelling request!

**Comment by jcouv** [comment]
Talked to Cyrus. I'll go ahead and start a championed proposal. Link coming shortly.
He also pointed out that there is some impact on an IDE feature that suggests adding named-parameters. It could just suggest a minimal fix (ie. not add names to all the rest of the parameters).

**Comment by jcouv** [comment]
Opened https://github.com/dotnet/csharplang/issues/570 to start the process. I'll write a proposal that can be discussed in LDM to get a rough sense of prioritization.

**Comment by svick** [comment]
@sharwell There are tools like Reviewable that build on top of GitHub's PRs. If you want to use Roslyn with PRs, I think an external tool like that is more likely than GitHub adding explicit support for C#.

**Comment by svick** [comment]
Would this allow using named arguments with `params`?

In https://github.com/dotnet/corefx/issues/19599#issuecomment-300527721, this method was proposed:

```c#
public class Task
{
    public static Task When(TaskStatus all, TaskStatus any, params Task[] tasks);
}
```

It would be great if such method could be called as:

```c#
Task.When(all: TaskStatus.RanToCompletion, any: TaskStatus.Faulted, task1, task2)
```

I think this would be useful for other `params` methods too.

**Comment by jcouv** [comment]
@sharwell The problem you point out is not new. Calling `Foo(3, 4)` picks one, while calling `Foo(3, x: 4)` picks the other. 
Granted this problem only occurs in trailing positions today.

```csharp
private static void Foo(int x, int y) { }
private static void Foo<T>(T y, int x) { }
```

**Comment by jcouv** [comment]
@svick Thanks for bringing up the `params` example. Yes, I think that should work (I took a note in [proposal ](https://github.com/dotnet/csharplang/blob/master/proposals/non-trailing-named-arguments.md) and I'll add test cases to the prototype). 

For each candidate method considered for applicability:
- If you use a named argument in the right position, the argument is used positionally. 
- If you use a named argument out-of-position, then following arguments must be named. Otherwise the candidate is not applicable.

In your example with `params`, the argument is used in the right position, so the rest of the parameters continue to work positionally.

**Comment by contextfree** [comment]
I also like to use named arguments whenever I call a method with multiple parameters of the same type; it makes it harder to accidentally get the order of parameters wrong, by "making wrong code look wrong". Currently this makes me also use named arguments for every other parameter as well, even when some of them have different types and so should be covered by the type checker.

**Comment by grzfes** [comment]
*(a bit late, but)*
Would not be good to give a warning for this known case (or suggestion):
```cs
private static void Foo(int x, int y) { }
private static void Foo<T>(T y, int x) { }

private static void Bar()
{
    Foo(3, 4);
    Foo(y: 3, 4); // warning
}
```

### Determining the problem
Collect candidates with an argument called `y`¹. Then compare overloads by type:
```
Foo(int, int)
Foo(T, int)
```
so as it usually happens: passing 2x `int` → `Foo(int, int)` is better.

### Message
Method `Foo<T>(T, int)` will be used instead of the preferred `Foo(int, int)` based on the positions of named arguments.

### Possible solutions

1. `Foo(int, int)` performs exactly the same as `Foo(T, int)` without a significant difference in performance (simple examples: `.ToString()`, `.GetHashCode()` etc.).

**Correction**: removal of unnecessary overload `Foo(int, int)`.

2. `Foo(int, int)` performs tasks much more efficiently for `int` (frequent use, less resource consumption, no casting etc.).

**Correction**:
```cs
private static void Foo(int x, int y) { }
private static void Foo<T>(int x, T y) { }
```
 or
```cs
private static void Foo(int y, int x) { }
private static void Foo<T>(T y, int x) { }
```
After change, a specialized overload `Foo(int, int)` will now be selected in both cases.

3. `T y`¹ means something different than `int y`¹.

**Correction**: change of argument name, e.g.:
```cs
private static void Foo(int z, int x) { }
private static void Foo<T>(T y, int x) { }

```
4. Inability to make changes 1-3.

**Correction**:
  * use of named parameters in the order defined in the preferred overload.
  * not to use named parameters for arguments that may cause this problem.
  * abandoning the use of named parameters for this method and its overloads.


*(not solution, but rather conscious use)*
Ignore the warning (suppress), if this is intentional call to the generic overload throughout positions of named arguments.


¹ - name of the first named argument that caused this problem.

<hr/>

I think that the unconscious changes in the way of choosing overload is a serious problem.
It is like suddenly there was no overloads like `Foo(int, int)`.

**Comment by sharwell** [comment]
@gafter Can we close this as now implemented?

**Comment by gafter** [comment]
Yes; this was implemented in C# 7.2 per https://github.com/dotnet/csharplang/issues/570
